feat: pass swagger-ui additional configuration

giovanni-bertoncelli opened the pull request:
<!--
Thank you for your pull request. Please provide a description above and review
the requirements below.

Bug fixes and new features should include tests and possibly benchmarks.

Contributors guide: https://github.com/fastify/fastify/blob/master/CONTRIBUTING.md

By making a contribution to this project, I certify that:

* (a) The contribution was created in whole or in part by me and I
  have the right to submit it under the open source license
  indicated in the file; or

* (b) The contribution is based upon previous work that, to the best
  of my knowledge, is covered under an appropriate open source
  license and I have the right under that license to submit that
  work with modifications, whether created in whole or in part
  by me, under the same open source license (unless I am
  permitted to submit under a different license), as indicated
  in the file; or

* (c) The contribution was provided directly to me by some other
  person who certified (a), (b) or (c) and I have not modified
  it.

* (d) I understand and agree that this project and the contribution
  are public and that a record of the contribution (including all
  personal information I submit with it, including my sign-off) is
  maintained indefinitely and may be redistributed consistent with
  this project or the open source license(s) involved.
-->
## Porpouse

This PR allows to pass a dynamic configuration to the generated swagger UI. The option that should be used in the plugin configuration in order to accomplish that is `uiConfig`.

Configuration docs: https://github.com/swagger-api/swagger-ui/blob/master/docs/usage/configuration.md

## Implementation

Since there is an issue configuring `swagger-ui-dist` (that distributes only static files, see https://github.com/swagger-api/swagger-ui/issues/5710) without changing the content of the static assets, the only way to configure the UI with a dynamic external configuration is by editing the `index.html` file and the script within that loads the swagger ui in the page.

There were two approaches:
1. Intercepting the `index.html` request on the documentation routes and distributing an altered html file that load swagger UI with the configuration merged with the one passed by the user.
2. Editing the file `index.html` when the project gets prepared (`npm run prepare`) in order to retrieving a remote configuration fetched through an XHR request on a designated endpoint.

Since the index.html file was already edited in the prepare stage I went for the second approach. Now a new route under `GET documentation/uiConfig` gets declared and returns the final ui configuration and the script inside the index page fetches the configuration from that endpoint. Finally the page loads the swagger ui from the given configuration.

## Drawbacks

Since there is no way to serialize functions over HTTP and JSON there is no way to pass functions to the parameters of the `uiConfig`, so only string/object/number configuration keys will be effective. Neither the first alternative implementation allowed to do that (at least without writing the plain function in the HTML file). However I think this is a fair compromise.

## Related issues

- #228 
- [swagger-ui#5710](https://github.com/swagger-api/swagger-ui/issues/5710)

## Others

The `npm run test` didn't run all the test inside the folder `test/` since the regex in the command was working only at the first directory level. Now it runs all the test till the 3rd directory level.

#### Checklist

- [X] run `npm run test` and `npm run benchmark`
- [X] tests and/or benchmarks are included
- [X] documentation is changed or added
- [X] commit message and code follows the [Developer's Certification of Origin](https://github.com/fastify/.github/blob/master/CONTRIBUTING.md#developers-certificate-of-origin-11)
      and the [Code of conduct](https://github.com/fastify/.github/blob/master/CODE_OF_CONDUCT.md)

giovanni-bertoncelli commented:
The final configuration passed to the Swagger UI is composed by merging sequentially:

- The default swagger UI configuration 
- The user defined configuration
- The `{url, oauth2RedirectUrl}` configuration

Fdawgs reviewed (commented):
Commit message for 9d96c34 should be `test`, not `feat`

mcollina commented on test/route.js:
```diff
@@ -86,6 +86,40 @@ test('/documentation/json route', t => {
+
+    t.match(payload, uiConfig, 'Invalid uiConfig received')
+  })
+})
```
Can you please add a test with a valid uiConfig as well?

giovanni-bertoncelli replied:
The configuration passed in that test is actually valid. Maybe I have to change the test message?

mcollina replied:
Sorry about it, I misunderstood the assertion.

mcollina reviewed (commented):
Good work!

mcollina reviewed (approved):
lgtm

climba03003 commented on package.json:
```diff
@@ -6,8 +6,8 @@
     "prepublishOnly": "npm run prepare",
-    "test": "standard && tap --100 test/**/*.js && npm run typescript",
-    "test:ci": "standard && tap test/**/*.js --coverage-report=lcovonly && npm run typescript",
+    "test": "standard && tap --100 test/*.js test/**/*.js test/**/**/*.js && npm run typescript",
```
I think this change is not needed as `test/**/*.js` means any files under `test` folder.

giovanni-bertoncelli replied:
I used to think that too but instead running with the npm it does not consider deeper level files...
As a notice: in fact the coverage shown in the repository badge is currently 46% instead of 100%...

climba03003 replied:
I face the same issue when I do a clean install of packages. It seems like `tap` changes the behavior of glob.

HighSoftWare96 replied:
Yep... Apparently it is a bug on their side..

giovanni-bertoncelli replied:
Yeah, probably a bug on their side

climba03003 replied:
@giovanni-bertoncelli
I finally see the cause, it should be the problem with bash, npm or yarn.
Can you test if adding a quote help to resolve it? It is a little bit too long if we include all the nested folder case.
```
"test": "standard && tap --100 'test/**/*.js' && npm run typescript",
"test:ci": "standard && tap 'test/**/*.js' --coverage-report=lcovonly && npm run typescript",
```

giovanni-bertoncelli replied:
Should be fine now

climba03003 reviewed (approved):
LGTM

I get stuck in your option 1 when I see this feature request. My other consideration was CSP, but updating the `index.html` in `npm run prepare` is a good solution without affecting CSP.
Good Approach.

climba03003 commented:
Just a suggest, will you consider replacing the whole block of `<script></script>` in `index.html`? I think you already changed all the content inside it.
It is more simple to use a single replace.

HighSoftWare96 commented:
> Just a suggest, will you consider replacing the whole block of `<script></script>` in `index.html`? I think you already changed all the content inside it.
> It is more simple to use a single replace.

Yeah, can do :)

giovanni-bertoncelli commented:
> Just a suggest, will you consider replacing the whole block of `<script></script>` in `index.html`? I think you already changed all the content inside it.
> It is more simple to use a single replace.

Yeah, that can do 😄

mcollina commented:
@giovanni-bertoncelli so we are waiting for a bit more change.

giovanni-bertoncelli commented:
> @giovanni-bertoncelli so we are waiting for a bit more change.

Yes. I'll replace the <script> and the package json in the next days..

giovanni-bertoncelli commented:
Ok. I've:
- Fixed the package.json `npm run test` with quotes (as @climba03003 suggested)
- Replacing entirely the `<script>` tag (as @climba03003 suggested)
- Renamed the unit test 


_(Sorry for the double github account, they are both mine)_

climba03003 commented on lib/mode/dynamic.js:
```diff
@@ -16,7 +16,8 @@ module.exports = function (fastify, opts, done) {
   if (opts.exposeRoute === true) {
     const prefix = opts.routePrefix || '/documentation'
-    fastify.register(require('../routes'), { prefix })
+    const uiConfig = opts.uiConfig || {}
```
```suggestion
    const uiConfig = Object.assign({}, opts.uiConfig)
```

mcollina reviewed (approved):
lgtm

giovanni-bertoncelli commented:
Any chance that this will be released soon? :)

Eomm commented:
Releasing 👍

giovanni-bertoncelli commented:
Sorry, I've seen the release but downloading with `npm` the resulting `static/` folder contains the old content... 🤔

Eomm commented:
mmm the script was updated
https://unpkg.com/browse/fastify-swagger@4.1.0/lib/util/prepare-swagger-ui.js

and it seems the pre-publish did not run:

https://github.com/fastify/fastify-swagger/blob/9d87a5fdfdcf4d7ffda39d01d4f696d2354bbf12/package.json#L7

I can release a fix this evening CEST if someone can't do it earlier

giovanni-bertoncelli commented:
Ok @Eomm is there any error in the package.json definition then?

mcollina commented:
I've released v4.1.1, try that.

giovanni-bertoncelli commented:
Yeah, now it's fine 🤷🏻‍♂️
Thanks @mcollina
